storage/innobase/xtrabackup/src/file_utils.cc

```diff
@@ -0,0 +1,46 @@
+#include "common.h"
+#include "my_io.h"
+
+#if defined _WIN32 || defined __CYGWIN__ \
+    || defined __EMX__ || defined __MSDOS__ || defined __DJGPP__
+# define _IS_DRIVE_LETTER(C) \
+    (((C) >= 'A' && (C) <= 'Z') || ((C) >= 'a' && (C) <= 'z'))
+# define HAS_DEVICE(Filename) \
+    (_IS_DRIVE_LETTER ((Filename)[0]) && (Filename)[1] == FN_DEVCHAR)
+# define FILE_SYSTEM_PREFIX_LEN(Filename) (HAS_DEVICE (Filename) ? 2 : 0)
+#else
+# define FILE_SYSTEM_PREFIX_LEN(Filename) ((void) (Filename), 0)
+#endif
+
+const char *safer_name_suffix(char const *file_name, int* prefix_len_out) {
```

Comment:
Wouldn't it be easier/cleaner/safer?
```
const string safer_name_suffix(const string &file_name, size_t* prefix_len_out) {

    static const string dots("..");
    static const string slash("/");
    auto start = file_name.rfind(dots);

    start = string::npos == start ? 0 : start + dots.length();
    start = file_name.find_first_not_of(slash, start);
    *prefix_len_out = string::npos == start ? file_name.length() : start;

    return file_name.substr(*prefix_len_out);
}
```

Reply:
I agree. But, please, mind I based this code in [gnu tar](https://git.savannah.gnu.org/cgit/paxutils.git/tree/paxlib/names.c#n100) and that most of the code in v2.4, and by extension v8.0, is C. So, please, I'd rather keep the C version.

Reply:
OK, it follows xbstream style which is C in 99.99%. If @rahulmalik87 is OK with this, I'm OK as well. Just wanted to stress out that C++ implementation took me 5 minutes, while I spent 30 minutes analyzing C-version pointers :)